### ISSUE TITLE
doc/fixing JupyterLab paths

### ISSUE THREAD

**Issue by djacu** [tweag/jupyterWith#213]
There is an ongoing issue with jupyter lab and the paths it is able to see. Ideally, it would only see the paths we specify, but it currently is exposed to directories in user space, etc, and local. One of the current main issues this creates is the newly created jupyter lab environment defaults to the python environment used to build the environment itself which we do not desire. The information below hopes to provide insight into how jupyter establishes its environment and sets paths.

Running `jupyter --paths` from the resulting build shows the `config`, `data`, and `runtime` directories.

```shell
$ ./result/bin/jupyter --paths
config:
    /home/bakerdn/.jupyter
    /nix/store/kx6q1gz2mm6maasahhkq3p4srk2hhgf6-python3-3.9.13-env/etc/jupyter
    /usr/local/etc/jupyter
    /etc/jupyter
data:
    /nix/store/mnpg1gshfxys72yi26zgvvdqpypws578-example_ansible-jupyter-kernel
    /nix/store/dgh0p09awjwd7gzi17pf6mwxvlxic16h-example_ansible_stable-jupyter-kernel
    /nix/store/4k4vqxjny5nnfpym0r3ca8xizwphhxmv-example_bash-jupyter-kernel
    /nix/store/b030rlqd1clkas2di0rmppk5rhiwf3kx-example_c-jupyter-kernel
    /nix/store/k4rmbp7vv4rj01davjkzkwcnr83n5cak-example_ipython-jupyter-kernel
    /nix/store/cwiy2ml5qlyfyjzc35dqzln9d061k0dy-example_nix-jupyter-kernel
    /nix/store/mzf7h6p365wl1y9b515cqlxq98188cfp-example_rust-jupyter-kernel
    /home/bakerdn/.local/share/jupyter
    /nix/store/kx6q1gz2mm6maasahhkq3p4srk2hhgf6-python3-3.9.13-env/share/jupyter
    /usr/local/share/jupyter
    /usr/share/jupyter
runtime:
    /home/bakerdn/.local/share/jupyter/runtime
```

Inspecting the `jupyter-lab` script inside the resulting build also provides additional information. The `JUPYTER_PATH` env variable has been manually modified to add new lines for readability.

```bash
$ cat ./result/bin/jupyter-lab
#! /nix/store/zwjm0gln1vk7x1akpyz0yxjsd1yc46gi-bash-5.1-p16/bin/bash -e
export JUPYTERLAB_DIR='/nix/store/kx6q1gz2mm6maasahhkq3p4srk2hhgf6-python3-3.9.13-env/share/jupyter/lab'
export JUPYTER_PATH='
/nix/store/mnpg1gshfxys72yi26zgvvdqpypws578-example_ansible-jupyter-kernel:
/nix/store/dgh0p09awjwd7gzi17pf6mwxvlxic16h-example_ansible_stable-jupyter-kernel:
/nix/store/4k4vqxjny5nnfpym0r3ca8xizwphhxmv-example_bash-jupyter-kernel:
/nix/store/b030rlqd1clkas2di0rmppk5rhiwf3kx-example_c-jupyter-kernel:
/nix/store/k4rmbp7vv4rj01davjkzkwcnr83n5cak-example_ipython-jupyter-kernel:
/nix/store/cwiy2ml5qlyfyjzc35dqzln9d061k0dy-example_nix-jupyter-kernel:
/nix/store/mzf7h6p365wl1y9b515cqlxq98188cfp-example_rust-jupyter-kernel
'
exec -a "$0" "/nix/store/dznq34m8hwdmzysm35ak6r74lwkwql6g-wrapper-python3-3.9.13-env/bin/.jupyter-lab-wrapped"  "$@"
```

Looking at the `config` paths first, we see in the [documentation](https://docs.jupyter.org/en/latest/use/jupyter-directories.html#configuration-files) that jupyter looks for config files in the `~/.jupyter` directory by default. We can see that in the first line under `config` from `jupyter --paths`. I found that if we modify the flake.nix file so that the wrapper environment sets the `JUPYTER_CONFIG_DIR`, we can remove that path. Modifying the last few lines of the `mkJupyterlabInstance` function to the following removes `/home/bakerdn/.jupyter` from `config` and replaces it with `/nix/store/kx6q1gz2mm6maasahhkq3p4srk2hhgf6-python3-3.9.13-env/share/jupyter/lab`. The other paths are still there, including the other path to the python environment in the nix store, so we'll have to find some way to explicitly define the path we want and not allow any others.

```nix
pkgs.runCommand "wrapper-${jupyterlab.name}"
{nativeBuildInputs = [pkgs.makeWrapper];}
  ''
  mkdir -p $out/bin
  for i in ${jupyterlab}/bin/*; do
    filename=$(basename $i)
    ln -s ${jupyterlab}/bin/$filename $out/bin/$filename
    wrapProgram $out/bin/$filename \
      --set JUPYTERLAB_DIR ${jupyterlab}/share/jupyter/lab \
      --set JUPYTER_PATH ${kernelsString requestedKernels} \
      --set JUPYTER_CONFIG_DIR ${jupyterlab}/share/jupyter/lab
  done
'';
```

In the `data` section, the first 7 paths ending in `-kernel`, are nix store paths to the available kernels we created. I believe they are set by `JUPYTER_PATH`. Similar to the `config` section, if `JUPYTER_DATA_DIR` is unset, it will add `~/.local/share/jupyter/` to the `data` paths. Where the `/usr` paths come from is unknown at this point but we likely want to remove them going forward.

Finally, similar to `config` and `data`, `runtime` is set by `JUPYTER_RUNTIME_DIR`. If unset, the documentation says it could be set to `$XDG_RUNTIME_DIR/jupyter` or a `runtime/` subdirectory of the user's data directory specified by `{sys.prefix}/share/jupyter/` where `sys.prefix` comes from Python. Neither of these seem to be the case here and it somehow picked up a directory under my user.

**Comment by djacu** [comment]
Current state of `jupyter` specific paths. Running `jupyter --paths --debug` gives extra information about paths.

- [ ] `JUPYTER_PREFER_ENV_PATH` is not set. This prefers user-level paths over env-level paths. We aim to disable all unwanted user paths going forward. It is likely that we want to set this.
- [x] `JUPYTER_NO_CONFIG` is not set. We do _not_ want this set as it will not use the config directories we have specified.
- [x] `JUPYTER_CONFIG_PATH` is not set. We likely do not need to set this. We already have a config directory set with `JUPYTER_CONFIG_DIR`.
- [x] `JUPYTER_CONFIG_DIR` is set. We set this to override using the default user-level config directory (e.g. `/$HOME/.local`) with a path in the nix store where we will explicitly place config files.
- [ ] Python's `site.ENABLE_USER_SITE` is not True, so it does not add the Python site user directory '/$HOME/.local'. We likely want this variable set to `False` or `None`. The [site module documentation](https://docs.python.org/3/library/site.html#site.ENABLE_USER_SITE) specifies that we can emulate setting the value to `False` by setting the `PYTHONNOUSERSITE` environment variable. The [jupyter-core changelog](https://jupyter-core.readthedocs.io/en/stable/changelog.html#id5) mirrors this information. It is unclear at this time how to set the value to `None` if needed; [this thread](https://stackoverflow.com/questions/25584276/how-to-disable-site-enable-user-site-for-an-environment) might provide some leads.
- [x] `JUPYTER_PATH` is set. We have set this to all of the requested kernels.
- [x] `JUPYTER_DATA_DIR` is set. We have set this to a path in the nix store. It is uncertain at the moment what, if anything, will be placed here. It also had the side effect of overriding the default user-level data directory.
- [x] `JUPYTER_RUNTIME_DIR` is set to '/path-not-set', overriding the default runtime directory. However, we have an override in place that patches the `jupyter-core` `jupyter_runtime_dir` function effectively disabling the environment variables influence. The run time directory is set to `/$HOME/.local/share/jupyter/runtime` as the directory needs user read/write permissions and only stores ephemeral data related to running jupyter applications.

**Comment by djacu** [comment]
Current state of `jupyter lab` specific paths. `jupyter lab` has 3 dedicate paths which are listed below and each item is marked as complete if the path is under nix control and does not include any unwanted user or global paths. There may still be issues going forward with write permissions but for the time being we want control over which directories `jupyter` has access to.

The paths can be found by running `jupyter lab paths`.

- [x] Application directory: This can be set with the `JUPYTERLAB_DIR` environment variable. We currently have this set to `${jupyterlab}/share/jupyter/lab` where `jupyterlab` is the jupyter environment that we create with `poetry2nix` and then wrap for our outgoing jupyter environment.
- [ ] User Settings directory: JupyterLab will use the `config` paths of `jupyter`. Since we have set the `config` path using the `jupyterDir` derivation, the resulting output directory will be `${jupyterDir}/config/lab/user-settings`. However, no such folder currently exists and needs to be created.
- [ ] Workspaces directory: JupyterLab will use the `config` paths of `jupyter`. Since we have set the `config` path using the `jupyterDir` derivation, the resulting output directory will be `${jupyterDir}/config/lab/workspaces`. However, no such folder currently exists and needs to be created.
